File: fable/examples/typescript/05_token_counting.ts
// 05 · Re-baseline your token math.
//
// Fable's new tokenizer counts ~30% more tokens than Opus-tier for the same
// text. Old budgets and cost estimates are now wrong. Passing
// model="claude-fable-5" to countTokens returns counts under BOTH tokenizers.
//
// Run:  npx tsx 05_token_counting.ts
import { MODEL, client } from "./_shared.js";

const SAMPLE =
  "The quick brown fox jumps over the lazy dog. ".repeat(40) +
  "\n\nfunction add(a, b) {\n  return a + b;\n}\n";

const r = await client.messages.countTokens({
  model: MODEL,
  messages: [{ role: "user", content: SAMPLE }],
});

const fresh = r.input_tokens;
// Newer SDK field; cast through unknown in case your version predates it.
const prior = (r as unknown as { input_tokens_prior_tokenizer?: number })
  .input_tokens_prior_tokenizer;

console.log(`Fable tokenizer (what you're billed):   ${fresh}`);
if (prior) {
  console.log(`Prior-generation tokenizer (Opus-tier): ${prior}`);
  console.log(`Delta:                                  ${(((fresh / prior) - 1) * 100).toFixed(1)}%`);
} else {
  console.log("(Upgrade @anthropic-ai/sdk to see input_tokens_prior_tokenizer.)");
}

console.log(`\nEstimated input cost at $10/1M: $${((fresh / 1_000_000) * 10).toFixed(6)}`);

File: fable/examples/python/05_token_counting.py
"""05 · Re-baseline your token math.

Fable uses a new tokenizer: the same text counts ~30% more tokens than on
Opus-tier models. Budgets, cost estimates, and compaction triggers calibrated
on Opus are now wrong — and you can silently truncate output.

The fix is one call. When you pass `model="claude-fable-5"`, count_tokens
returns counts under BOTH tokenizers so you can see the delta on your own
prompts. Never apply a blanket multiplier.

Run:  python 05_token_counting.py
"""

from _shared import MODEL, client

SAMPLE = (
    "The quick brown fox jumps over the lazy dog. " * 40
    + "\n\ndef add(a, b):\n    return a + b\n"  # code tokenizes differently
)

r = client.messages.count_tokens(
    model=MODEL,
    messages=[{"role": "user", "content": SAMPLE}],
)

new = r.input_tokens
old = getattr(r, "input_tokens_prior_tokenizer", None)

print(f"Fable tokenizer (what you're billed):   {new}")
if old:
    print(f"Prior-generation tokenizer (Opus-tier): {old}")
    print(f"Delta:                                  {(new / old - 1) * 100:+.1f}%")
else:
    print("(Your SDK didn't return input_tokens_prior_tokenizer — upgrade the")
    print(" `anthropic` package to see the side-by-side comparison.)")

# Rough cost preview at Fable's input price ($10 / 1M tokens):
print(f"\nEstimated input cost at $10/1M: ${new / 1_000_000 * 10:.6f}")
